lighthouse-viewer/app/src/lighthouse-report-viewer.js

```diff
@@ -35,12 +35,13 @@ class LighthouseReportViewer {
 
     this._dragAndDropper = new DragAndDrop(this._onFileLoad);
     this._github = new GithubApi();
+    this._psi = new PSIApi();
 
     /**
-     * Used for tracking whether to offer to upload as a gist.
-     * @type {boolean}
+     * Used for tracking where the report came from.
+     * @type {'gist' | 'psi' | 'opener' | 'paste' | 'file' | null}
```

Comment:
this seems like a YAGNI situation. We only need a single bit for if we should clear the query string or not, and it's only used in the rare case that a user starts with a report loaded from a gist or psi and then decides to load a different one from a local file or a c/p.

`clearQueryStringOnReplace` or whatever seems fine, and probably would allow a bunch of deletions...just needs to be set to false on gist and psi load and have `_replaceReportHtml` always set it back to true at the end of a load.

Reply:
It's not just one bit. did you see `saveCallback`?

Reply:
> It's not just one bit. did you see saveCallback?

good point. This is probably just going to be ugly because we're trying to take a very simple custom router and expand it without rewriting it :) (see e.g. `_loadFromGistURL`)

maybe the best thing is to just go back to the relative simplicity of `_reportIsFromGist`/`_reportIsFromPsi`, sorry :(

Reply:
yo these booleans are straight trash. 🗑   :)
since loading a report only one way at a time, i think its way better to have this enum.

Reply:
As a middle ground, @brendankenny, what if I removed all the enum values which we don't care about?

Reply:
> yo these booleans are straight trash. 🗑 :)
> since loading a report only one way at a time, i think its way better to have this enum.

I would agree with you if there was any reason to have this information. Why architect a system that nothing uses?

Reply:
![image](https://user-images.githubusercontent.com/4071474/63191953-b59c9300-c01e-11e9-8c90-2c019e5f9bc0.png)

sooooo what shall it be then?

Reply:
trying to sum up what I was saying on chat:

- if we have the full enum of possible sources we do a bunch of bookkeeping for no purpose since nothing ever reads from that variable except where we check whether the window hash should be updated and if we should have "save to gist" in the dropdown.
- If we had an enum of `'gist'|'psi'|null` we limit to actually used information, but when we start doing the redirect to web.dev and delete the `psi` state, we either have `'gist'|null` or someone has to convert back to a boolean.

If on the other hand we use the two boolean state, it's kind of ugly now (as @paulirish notes, we have three possible app states but four possible variable states) but it's just a delete of the single `_reportIsFromPsi` to get us to a clean slate when the web.dev redirect happens.

This PR does make it apparent that this is a bad variable, conflating report-is-from-gist and report-*will-be*-from-gist, and it just happens to work because the code never cares about the report-is-from-gist case (well it's bad in multiple ways, but that's the most pertinent right now :). Really they're only used for the case of report-will-be-from-gist(-as-soon-as-_replaceReportHtml-runs).

That actually points the way to make this simpler, as well. Keep two booleans, defaulted to `false`, set them both back to `false` at the end of `_replaceReportHtml` (in a `finally` block, or whatever). Then one or the other needs to be set to `true` in either the gist or psi cases in `loadFromDeepLink`. No need to have `true` or `false` sprinkled elsewhere.